strings/base_collections_vector.h

```diff
@@ -100,6 +100,14 @@ namespace winrt::impl
 
         bool IndexOf(Windows::Foundation::IInspectable const& value, uint32_t& index) const
         {
+            if constexpr (!std::is_base_of_v<Windows::Foundation::IUnknown, T>)
+            {
+                if (value == nullptr)
+                {
+                    return false;
+                }
+            }
+
             return IndexOf(unbox_value<T>(value), index);
```

Comment:
This will still crash if somebody searches for, say, an integer in an `IVector<hstring>`. Should return "not found".

I was thinking more along the lines of

```cpp
template<typename T>
std::optional<T> try_unbox(Windows::Foundation::IInspectable const& value)
{
        if (value)
        {
            if constexpr (std::is_base_of_v<Windows::Foundation::IUnknown, T>)
            {
                if (auto temp = value.try_as<T>())
                {
                    return temp;
                }
            }
            else if constexpr (std::is_enum_v<T>)
            {
                if (auto temp = value.try_as<Windows::Foundation::IReference<T>>())
                {
                    return temp.Value();
                }

                if (auto temp = value.try_as<Windows::Foundation::IReference<std::underlying_type_t<T>>>())
                {
                    return static_cast<T>(temp.Value());
                }
            }
            else
            {
                if (auto temp = value.try_as<Windows::Foundation::IReference<T>>())
                {
                    return temp.Value();
                }
            }
            return {};
        }
        else
        {
            // nullptr is valid for reference types.
            if constexpr (std::is_base_of_v<Windows::Foundation::IUnknown, T>)
            {
                return T{nullptr};
            }
            else
            {
                // nullptr is not valid for value types.
                return {};
            }
        }
}
```

And then we have

```cpp
bool IndexOf(Windows::Foundation::IInspectable const& value, uint32_t index) const
{
    auto unboxed = try_unbox<T>(value);
    if (!unboxed) return false; // wrong type, cannot possible be in the vector
    return IndexOf(unboxed.value(), index);
}

```

Reply:
(I think `try_unbox` would be generally useful.)

Reply:
Yeah that seems useful.